### ISSUE TITLE
adding ArcGIS JS API to the app..? 

### ISSUE THREAD

**Issue by Deasus** [thkruz/keeptrack.space#661]
Im really impressed with the work youve done with KeepTrack! -awesome job. I had a quick question: would it be possible to add the ArcGIS Javascript API to the app to leverage a mapped globe instead of the current image of the earth being used..?
Im asking because you could add layers to the globe for ground-based map layers etc. 

I assume the process could be similar to: https://developers.arcgis.com/javascript/latest/sample-code/satellites-3d/

**Comment by thkruz** [comment]
Sorry for the delay getting back to you (saw your email).

Are you asking if it is technically feasible or if I will personally implement it?

From a technical standpoint, the earth.ts file has a draw loop that puts out the earth object. It should be as simple as replacing that with the arcgis code like you suggested. The one issue you will run into is getting the earth to point in the right direction. KeepTrack draws everything in ECI, so as the earth rotates, the earth object has to be rotated. Many other tools treat the world as ECF so that the earth doesn't have to rotate and the satellites can just be adjusted instead. I am not sure how the arcgis code works.

In terms of me personally implementing it - I won't say no, but I will say it is a low priority. If I am not mistaken, pulling the layers for that will require an internet connection. KeepTrack is designed to be able to be run completely offline. I am open to exploring the idea and maybe implementing it with a flag in the settings so that you can choose to use it when online.

**Comment by Deasus** [comment]
Thanks for the response Theo! ArcGIS has been approved for quite some time in AFRL and USSF (as they brought over practices from the USAF). As a result they have quite a few earth layers for their “arc scenes”. This includes real time locations of ships, aircraft and launches. It would be interesting to see it Implemented. I’ve seen a number of people that have done so at a very high level… but not in as much detail as what you’ve done (awesome job by the way)! Additionally, the ArcGIS JS Api makes use of the Satellite.js library and as a result/need after load time, can work off line. It also can be configured to cache and work offline too. Let me know where the rath look is and I can try toReplace it to see what happens… also I could set the rotation of the globe if it doesn’t auto adjust.Let me know your thoughts, Deepinder On Apr 1, 2023, at 11:26 AM, Theodore Kruczek ***@***.***> wrote:﻿
Sorry for the delay getting back to you (saw your email).
Are you asking if it is technically feasible or if I will personally implement it?
From a technical standpoint, the earth.ts file has a draw loop that puts out the earth object. It should be as simple as replacing that with the arcgis code like you suggested. The one issue you will run into is getting the earth to point in the right direction. KeepTrack draws everything in ECI, so as the earth rotates, the earth object has to be rotated. Many other tools treat the world as ECF so that the earth doesn't have to rotate and the satellites can just be adjusted instead. I am not sure how the arcgis code works.
In terms of me personally implementing it - I won't say no, but I will say it is a low priority. If I am not mistaken, pulling the layers for that will require an internet connection. KeepTrack is designed to be able to be run completely offline. I am open to exploring the idea and maybe implementing it with a flag in the settings so that you can choose to use it when online.

**Comment by thkruz** [comment]
Few clarifications. ArcGIS Api does not make use of satellite.js. Their example simply imports it. Since I am one of the main maintainers on that library, I would highly recommend that you ensure you are using the latest version - I recently corrected some math errors that impacted a small subset of the catalog. Alternatively, I have put out [ootk](https://github.com/thkruz/ootk) which is an expanded version of satellite.js in TypeScript.

While I am aware that it can run offline once it is done downloading, for me to replace the earth with ArcGIS, I need a layer that can be hosted locally versus pulling from an intranet (SIPR, etc).

Earth is located here: https://github.com/thkruz/keeptrack.space/blob/develop/src/js/drawManager/sceneManager/earth.ts

I am taking a look at it right now.

**Comment by thkruz** [comment]
Turns out this is a non-trivial project. I don't see a way to add an ArcGIS object into another canvas. Someone would need to migrate all of the non-earth objects of KeepTrack into the ArcGIS SceneView. That also means you would lose control over a lot of the drawing loop which is how I have optimized it to handle such a large number of objects at once.

Happy to answer specific questions if you go after this, but I am going to mark this as closed because I am not going to go after this one. If you have any specific use case you are trying solve, let me know and I will see if there is a compromise that will work.
